### PR TITLE
FPO structure + LayerZero module

### PR DESCRIPTION
**Pull request by FranklinWaller**

### CODE REVIEW

File: src/models/AppConfig.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    appConfig.networks = config.networks.map((networkConfig) => {
+        const parsedNetworkConfig = parseUnparsedNetworkConfig(networkConfig);
+        const network = AVAILABLE_NETWORKS.find(network => network.type === parsedNetworkConfig.type);
+
+        if (!network) throw new Error(`Network type "${parsedNetworkConfig.type}" does not exist`);
+
+        return new network(parsedNetworkConfig);
+    });
+
+    if (!config.modules || !Array.isArray(config.modules)) throw new Error(`atleast 1 item in "modules" is required and it must be an array`);
```
Comment (mariocao) [inline src/models/AppConfig.ts:35]:
```suggestion
    if (!config.modules || !Array.isArray(config.modules)) throw new Error(`at least 1 item in "modules" is required and it must be an array`);
```

File: src/models/AppConfig.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    appConfig.modules = config.modules.map((moduleConfig) => {
+        const parsedModuleConfig = parseUnparsedModuleConfig(moduleConfig);
+        const listener = AVAILABLE_LISTENERS.find(listener => listener.type === parsedModuleConfig.type);
+
+        if (!listener) throw new Error(`Listener type "${parsedModuleConfig.type}" does not exist`);
+
+        return new listener(parsedModuleConfig, appConfig);
+    });
+
+
```
Comment (mariocao) [inline src/models/AppConfig.ts:46]: remove :)

File: src/models/Block.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
+import Big from "big.js";
+
+export interface Block {
+    hash: string;
+    receiptRoot: string;
+    number: Big;
+}
+
+
+
```
Comment (mariocao) [inline src/models/Block.ts:10]:
```suggestion
```

File: src/models/AppConfig.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
```
Comment (mariocao) [inline src/models/AppConfig.ts:47]:
```suggestion

```

File: src/models/Block.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    type: 'hash';
+}
+
+export interface BlockTagType {
+    tag: string;
+    type: 'tag';
+}
+
+export type BlockTag = BlockHashType | BlockTagType;
+
```
Comment (mariocao) [inline src/models/Block.ts:22]:
```suggestion
```

File: src/models/Network.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+    constructor(config: NetworkConfig) {
+        super();
+        this.networkConfig = config;
+        this.id = `${config.type}-${config.networkId}`;
+        this.queue = new Queue(this.id);
+        this.networkId = config.networkId;
+    }
+
+    async onQeueuBatch(batch: DataRequestBatch) {
```
Comment (mariocao) [inline src/models/Network.ts:48]:
```suggestion
    async onQueueBatch(batch: DataRequestBatch) {
```

**Comment by FranklinWaller** [comment]
Continuing PR on #2

**Pull request closed without merging**